packages/lib-web/lib/src/conv.rs

```diff
@@ -0,0 +1,177 @@
+use js_sys::{Array, ArrayBuffer, Object, Reflect, Uint8Array};
+use serde::ser::Serialize;
+use serde_json::Value;
+use wasm_bindgen::{prelude::*, JsCast};
+use web_sys::Blob;
+
+extern crate base64;
+
+type Path = Vec<String>;
+type Blobs = Vec<(Path, Vec<u8>)>;
+
+#[inline(always)]
+fn js_to_serde_value(obj: JsValue) -> Value {
+    serde_wasm_bindgen::from_value(obj).unwrap()
+}
+
+#[inline(always)]
+fn serde_to_js_value(value: &Value) -> JsValue {
+    let serializer = serde_wasm_bindgen::Serializer::new().serialize_maps_as_objects(true);
+    value.serialize(&serializer).unwrap()
+}
+
+#[inline(always)]
+fn json_str_to_serde_value(s: &str) -> Value {
+    serde_json::from_str(s).unwrap()
+}
+
+#[inline(always)]
+fn serde_value_to_json_string(value: &Value) -> String {
+    value.to_string()
+}
+
+fn make_blob(bytes: &[u8]) -> Blob {
+    let ta = unsafe { Uint8Array::view(bytes) };
+    let arr = Array::new_with_length(1);
+    arr.set(0, JsValue::from(ta));
+    Blob::new_with_u8_array_sequence(&arr).unwrap()
+}
+
+fn read_array_buffer(ab: &JsValue) -> Vec<u8> {
+    Uint8Array::new(ab).to_vec()
+}
+
+fn replace_strings_with_placeholders(value: &mut Value) -> Blobs {
+    let mut blobs: Blobs = vec![];
+    replace_strings_with_placeholders_recursive(value, &vec![], &mut blobs);
+    blobs
+}
+
+fn replace_strings_with_placeholders_recursive(value: &mut Value, path: &Path, blobs: &mut Blobs) {
+    match value {
+        Value::Object(obj) => {
+            for (key, value) in obj.iter_mut() {
+                let mut newpath = path.clone();
+                newpath.push(key.clone());
+                replace_strings_with_placeholders_recursive(value, &newpath, blobs)
+            }
+        }
+        Value::String(s) => {
+            let bytes = base64::decode(s).unwrap();
+            blobs.push((path.clone(), bytes));
+            *value = Value::Null
+        }
+        _ => (),
+    }
+}
+
+fn replace_placeholders_with_blobs(root: &JsValue, blobs: &Blobs) {
+    for (path, bytes) in blobs.iter() {
+        replace_placeholder_with_blob(root, path, bytes);
+    }
+}
+
+fn replace_placeholder_with_blob(root: &JsValue, path: &[String], bytes: &[u8]) {
+    match path {
+        [head] => {
+            let key = head.into();
+            let blob = make_blob(bytes);
+            unsafe { Reflect::set(&root, &key, &blob) }.unwrap();
+        }
+        [head, ..] => {
+            let key = head.into();
+            let child = unsafe { Reflect::get(&root, &key) }.unwrap();
+            replace_placeholder_with_blob(&child, &path[1..], bytes);
+        }
+        [] => panic!("Empty path"),
+    }
+}
+
+pub fn parse(s: &str, return_blob: bool) -> JsValue {
+    let mut value = json_str_to_serde_value(s);
+    let blobs = if return_blob {
+        replace_strings_with_placeholders(&mut value)
+    } else {
+        vec![]
+    };
+    let jsvalue = serde_to_js_value(&value);
+    replace_placeholders_with_blobs(&jsvalue, &blobs);
```

Comment:
I don't know Rust much, and just wonder if it's possible to convert binaries into blobs (if needed) in a single run when iterating through the value? (instead of converting them to bytes first and then making blobs)

Reply:
Another solution would be to extend `serde_json::value::Value` with our binary pointer type (basically an object wrapping `ptr` and `len` fields), then detect such pointers while parsing JSON and finally implement a trait like `Into<JsValue>` which would be invoked automatically during conversion to JS object. But let's wait until the pointer optimization is ready as the current version is much simpler as well as easier to debug and profile. Once it's finished, I'll try to find a way to handle it nicely :)

Reply:
I see, good!